src/syscalls/business_logic_syscall_handler.rs

```diff
@@ -944,4 +955,65 @@ impl<'a, S: StateReader> BusinessLogicSyscallHandler<'a, S> {
             body: None,
         })
     }
+
+    fn keccak(
+        &mut self,
+        vm: &mut VirtualMachine,
+        request: KeccakRequest,
+        remaining_gas: u128,
+    ) -> Result<SyscallResponse, SyscallHandlerError> {
+        let length = (request.input_end - request.input_start)?;
+        let mut gas = remaining_gas;
+
+        if length % 17 != 0 {
+            let response = self.failure_from_error_msg(vm, b"Invalid keccak input size")?;
+            return Ok(SyscallResponse {
+                gas,
+                body: Some(response),
+            });
+        }
+        let n_chunks = length / 17;
+        let mut state = [0u64; 25];
+        for i in 0..n_chunks {
+            // TODO: check this before the loop, taking care to preserve functionality.
+            if gas < KECCAK_ROUND_COST {
+                let response = self.failure_from_error_msg(vm, b"Syscall out of gas")?;
+                return Ok(SyscallResponse {
+                    gas,
+                    body: Some(response),
+                });
+            }
+            gas -= KECCAK_ROUND_COST;
+            let chunk_start = (request.input_start + i * 17)?;
+            let chunk = get_felt_range(vm, chunk_start, (chunk_start + 17)?)?;
+            for (i, val) in chunk.iter().enumerate() {
+                state[i] ^= val.to_u64().unwrap();
```

Comment:
Can we guarantee `val` will be in range for conversion to `u64`? Otherwise, convert this to an error, since someone could use it to crash the node.